# <span style="color: #FF0000" >SNT - INITIATION ET REVISIONS EN LANGAGE PYTHON - CORRECTION</span>

# <span style="color: #008F00" > 1. Présentation du Notebook Jupyter</span>

Il s'agit du document que vous avez sous les yeux, document au format .ipynb. 
Sans rentrer dans les détails, un **notebook** fonctionne avec un système de cases qui peuvent correspondre soit à du texte soit à des programmes exécutables, c'est cette dernière fonctionnalité qui est utile pour nous car cela vous évitera d'avoir à installer un programmae supplémentaire.

Vous êtes ici dans une zone de texte (lisible mais aussi est modifiable).
Pour la modifier, double cliquer dessus, réalisez vos modifications. Lorsque vous avez terminé, il suffit d'appuyer sur la touche shift (*la flèche vers le haut à gauche du clavier*) et en même temps sur entrée pour repasser à la visions normale.

Lorsque la case comme celle ci-dessous est exécutable (un programme en langage python), il suffit d'appuyer sur la touche shift (*la flèche vers le haut à gauche du clavier*) puis entrée pour exécutez le programme.
Testez l'exemple ci-dessous. En dessous de la cellule s'affichera le résultat de l'exécution du code.

In [ ]:
print("Cellule exécutable")
a=3
print(a)

Il existe plusieurs tutoriels sur internet, les notebook jupyter sont mêmes utilisables depuis lycée connecté.
Par exemple [ici](http://python.lecoinduprogrammeur.org/2017/05/07/jupyter-notebook-ecrivez-executer-documentez-et-publiez-votre-code-python/)
Si vous préférez, vous pouvez tout à fait installer `spyder` (que nous utilisons au lycée) ou un éditeur en ligne comme [repl.it](https://repl.it/languages/Python3).

# <span style="color: #008F00" >2. Quelques révisions</span>

La plupart d'entre vous ont vu au collège la programmation avec scratch : commencez par lire le livre page 172 et 173 puis faire ce petit [QCM de révision :](https://numeres.net/9782206103389/res/9782206103389-ht5-snt-15/index.html).  
Le corrigé est disponible page 188.  

In [ ]:
# Shift + entrée pour exécuter la cellule, ici lire la vidéo.
from IPython.display import Video
Video("http://physiquelga.free.fr/SNT/Programmation/PythonHistoire.mp4")

# <span style="color: #008F00" >3 Variables et instructions élémentaires :</span>
Les variables permettent d'enregistrer, de stocker des informations. L'intérêt peut être la clarté de lecture du programme (par exemple entre utiliser le nombre 23 et dire nb_eleves = 23, la différence est que l'on sait de quoi on parle). La variable permet surtout de conserver une information que l'on réutilisera par la suite. Les fonctions input() et print() seront étudiées ici.
**Lire les pages 174, 175 de votre livre.**

In [ ]:
# Lorsque le signe # précède un bout de code, il s'agit d'un commentaire
# Le commentaire aide à comprendre le programme mais n'en fait pas partie.
# Apprendre à écrire un programme page 175
R=int(input("Rouge="))
V=int(input("Vert="))
B=int(input("Bleu="))
G=int(0.11*R+0.83*V+0.06*B)
print("G=",G)

## 3.1 Les réseaux sociaux (n°1 page 174) :
Un réseau social gagne chaque mois 50 euros par abonné et dépense 1500 euros. On veut calculer le revenu annuel du réseau social en fonction du nombre d'abonnés.
<img src="http://physiquelga.free.fr/SNT/Programmation/3RS.jpg" width=35%>

In [1]:
# Déterminer la ou les variables à saisir, le traitement et l'affichage à réaliser.
# Ecrire le programme correspondant en Python.
# Le nombre d'abonné sera à fournir par l'utilisateur.
# Il y a deux variables, deux entiers "nbabonne" et "revenu".
nbabonne = int(input("Combien d'abonnés avez-vous ?"))
revenu = nbabonne * 50 - 1500
print ("Le revenu du réseau social en euros est de ",end="")
print(revenu)

Combien d'abonnés avez-vous ? 25


Le revenu du réseau social en euros est de -250


## 3.2 La localisation
Un centimètre sur une carte représente 2 kilomètres dans la réalité. On souhaite traduire la longueur mesurée sur la carte en longueur réelle.  
<img src="http://physiquelga.free.fr/SNT/Programmation/3L.jpg" width=35%>

In [ ]:
# Déterminer la ou les variables à saisir, le traitement et l’affichage à réaliser.
# Écrire le programme correspondant en Python.
longueurMesuree = int(input("Quelle distance mesurez-vous en cm?"))
longueurReelle = longueurMesuree * 2
print(f"Dans la réalité, la distance est de {longueurReelle} km.")

Pour afficher le texte, j'ai ici utilisé une f-string. Pour comprendre commet cela marche, [regarder ici](https://he-arc.github.io/livre-python/fstrings/index.html).

## 3.3 Exercices

**N°12 page 182 :**
Quel est le type de chaque variable x, y, z et t ? 

>```python
x=2.5
y="Bonjour"
z=7
t="18"
```
    

In [ ]:
#exemple : 
x = 2.5
print(type(x))
y="Bonjour"
print(type(y))
z=7
print(type(z))
t="18"
print(type(t))

Nous avons donc un nombre réel (float) un entier et deux chaînes de caractère.

**N°13 page 182 :** Qu’affiche le programme ?

>```python
x=3
x=5
x=2
print(x)
```

In [ ]:
x=3
x=5
x=2
print(x)

**N°17 page 183 :** Juliette a écrit le programme suivant pour écrire les multiples de 25 :

>```python
x=input("Donner la valeur de x")
y=25
print(x*y)

1. Quel est le type de la variable x dans ce programme ?
2. Trouver l’erreur.
3. Corriger le programme.

In [ ]:
# x est une chaîne de caractère
# Il faut que le type soit en entier...
x=int(input("Donner la valeur de x"))
y=25
print(x*y)

# <span style="color: #008F00" >4 Instructions conditionnelles :</span>
Les instructions conditionnelles sont un éléemnt essentiel en programmation qui va permettre au programme de réagir différemment selon les données qu'il reçoit. En fançais, si... alors ...
La notion d'indentation sera ici abordée.
**Lire les page 176 et 177 de votre livre.**

In [ ]:
age = int(input("Quel est votre âge ?"))
if age<13 :
    print("Inscription interdite")
elif age<15:
    print("Autorisation nécessaire")
else:
    print ("Inscription libre")

## 4.1 La photographie numérique :
Afin de repérer la couleur dominante parmi les valeurs de rouge (R), de vert (V) et de bleu (B) d’une couleur, on souhaite afficher quelle est la plus grande valeur parmi les trois.
<img src="http://physiquelga.free.fr/SNT/Programmation/4PN.jpg" width=35%>

In [ ]:
# Déterminer le nombre de conditions et les écrire.
# Écrire le programme correspondant en Python.
# Il faut tester les 3 couleurs, on peut aussi tester des combinaisons de test avec les mot clé "and" et "or".
R=int(input("Rouge="))
V=int(input("Vert="))
B=int(input("Bleu="))
if R>=V :
    if R>= B :
        print ("Le rouge domine")
    else :
        print("Le bleu domine")
elif V>=B :
    print ("Le vert domine")
else : 
    print("Le bleu domine")

## 4.2 Exercices

**N°23 page 183 :**
Écrire en Python la suite d’instruction qui affiche une alerte si la température (on utilisera la variable « temperature ») dépasse 40 °C.

In [ ]:
temperature = int(input("Quelle est la température ?"))
if temperature >= 40 :
    print("Alerte !")

**N°25 page 182 :**
On donne le programme suivant :

1. Qu’affiche ce programme lorsque x = 150, puis x = 300 ?
2. On remplace la deuxième instruction « if »  par l’instruction « elif ».Qu’affiche le nouveau programme lorsque x = 150, puis x = 300 ?

In [ ]:
x = int(input())
if x>100 :
    print("x est grand")
if x>200:  
    print("x est très grand")
else:  
    print("x est petit")

In [ ]:
x = int(input())
if x>100 :
    print("x est grand")
elif x>200:  
    print("x est très grand")
else:  
    print("x est petit")

# <span style="color: #008F00" >5 Les boucles :</span>
Ces structures permettent de répéter des actions ce qui bien évidemment arrive très souvent dans un programme. Il existe plusieurs types de boucles dont notamment les boucles itératives ou boucles bornées : on sait à l'avance le nombre de fois où la boucle se fera (mot clé `for`). Il existe également les boucles non-bornées qui se feront aussi longtemps qu'une condition sera valable (mot clé `while`). **Lire page 178 et 179 de votre livre.**

## 5.1 Les réseaux sociaux (n°5 page 179) :
Le nombre d’inscrits sur un nouveau réseau social est de 400 000 en janvier. Il est multiplié par 1,1 tous les mois.On souhaite écrire un programme estimant au bout de combien de mois il y aura plus d’1 million d’inscrits.  
<img src="http://physiquelga.free.fr/SNT/Programmation/5RS.jpg" width=35%>

In [ ]:
# Déterminer le type de boucle nécessaire ainsi que l’instruction à répéter : on ne connaît pas à l'avance le nombre de mois donc boucle while...
# Écrire le programme correspondant en Python.
nbinscrit = 400000
nbmois = 0
while nbinscrit <= 1000000 :
    nbinscrit = nbinscrit * 1.1
    nbmois = nbmois + 1
print (nbmois)

## 5.2 Exercices

**N°27 page 182 :** Qu’affiche ce programme ?
>```python
for i in range(10): 
    print(i*i)

# Il affiche les 10 premiers carrés.

In [ ]:
for i in range(10): 
    print(i*i)

**N°31 page 183 :** Combien de tours effectue cette boucle non bornée ?

>```python
x=0
while x<10:
    x=x+3
```

In [ ]:
# Elle effectue 4 tours 
x=0
while x<10:
    x=x+3
    print(x)

**N°34 page 183 :** Recopier et compléter le programme qui affiche les heures et les minutes d’une journée de 0 h 00 à 24 h 00.
>```python
for i in …:
    for j in …:
        print(…,"h",…,"min")
```

In [ ]:
for i in range(24):
    for j in  range(60):
        print(i,"h",j,"min")

**N°33 page 183 :** Recopier et compléter le programme Python qui affiche « voulez-vous continuer ? » tant que l’utilisateur répond « oui ».
```
rep="oui"
… 
    rep=input("Voulez-vous continuer ?")```

In [ ]:
rep="oui"
while rep == "oui" :
    rep=input("Voulez-vous continuer ?")
print ("On s'arrête...")

# <span style="color: #008F00" >6 Les fonctions :</span>
Si en théorie, les fonctions ne sont pas obligatoires pour coder, elles se révèlent vites indispensables. Que ce soit pour rendre votre code plus lisible, pour travailler à plusieurs, pour réutiliser des morceaux de code, il faut savoir écrire une fonction. **Lire les pages 180 et  181 de votre livre.**

Fiche méthode page 181 : pour automatiser un changement d’échelle de carte, on veut réaliser une fonction qui calcule la longueur sur la carte après le changement d’échelle. Si on note **e1 l’échelle initiale** et **d la longueur initiale**, la distance réelle est égale à d / e1. La longueur L est égale à distanceRéelle x e2 **e2, l'échelle finale**.
Écrire une fonction qui donne la longueur finale calculée à partir des échelles de départ et d’arrivée et de la longueur initiale.
Par exemple, si l'échelle e1 est 1/25000 (1 cm pour 250 m), e2 est 1/100000 (1cm pour 1 km), une longueur de 8 cm sur la carte 1 représentera seulement 2 cm sur la carte 2.

In [ ]:
def changementDechelle(d,e1,e2):
    distanceReelle=d/e1
    L=distanceReelle*e2
    return L

# Pour appeler la fonction, vous pouvez essayer d'autres échelle et d'autres valeurs.
changementDechelle(8,1/25000,1/100000)

## 6.1 Les données :
L’abonnement à un service de stockage de données en ligne pour une entreprise coûte 200 euros d’ouverture de compte à l’inscription, puis 340 euros par mois. On veut calculer le prix total payé par l’entreprise en fonction du nombre de mois d’abonnement.
<p style ="text-align = center"><img src="http://physiquelga.free.fr/SNT/Programmation/6D.jpg" width=35%></p>

In [ ]:
"""" Déterminer les paramètres : le nombre de mois, 
la valeur de sortie : c'est le prix total et les instructions : faire le calcul.""" 
# Écrire une fonction en Python calculant le prix total payé par l’entreprise.
def prixTotal(nbmois) :
    prix = 200 + 340*nbmois
    return (prix)

In [ ]:
# Pour tester la fonction, inidquer le nombre de mois à l'intérieur des parenthèses.
prixTotal(...)

## 6.2 La photographie numérique
On souhaite calculer la résolution R d’une photo en fonction de sa définition D et de ses dimensions l et L. On donne : R = D / (l x L)
<p style ="text-align = center"><img src="http://physiquelga.free.fr/SNT/Programmation/6PN.jpg" width=35%></p>

In [ ]:
"""" Déterminer les paramètres : il y en a 3, la valeur de sortie est la résolution.
Les instructions sont données par l'énoncé)."""
# Écrire la fonction en Python calculant la résolution.
def calculResolution(definition,largeur,longueur) :
    res = definition / (largeur*longueur)
    return(res)

In [ ]:
#Les valeurs sont ici celle d'une télé HD de 102 cm de diagonale
calculResolution(1920*1080,49,89)